File: 0037-sudoku-solver/0037-sudoku-solver.cpp
// Dedicated to Junko F. Didi and Shree DR.MDD  

#include <bits/stdc++.h>
using namespace std;

class Solution {
	unsigned rowMask[9]={0};
    unsigned colMask[9]={0};
    unsigned boxMask[9]={0};
    vector<pair<char, char>> blanks;
public:
     void print(vector<vector<char>>& mat){
        for(vector<char>& r: mat){
            for(char& cc: r)
                cout<<cc;
            cout<<endl;
        }
        cout<<"===========\n";
    }
    void applyMask(int r, int c,  int val){
        const int bit=1<<val;
		rowMask[r]|=bit;
        colMask[c]|=bit;
        const int boxId=(r/3)*3 +c/3;
        boxMask[boxId]|=bit;
	}

    void initialize(vector<vector<char>>& mat) {
        blanks.reserve(81);
        for (int r=0; r<9; r++) {
            for (int c=0; c<9; c++) {
                char cc=mat[r][c];
                if (cc=='.') {
                    blanks.emplace_back(r, c);
                }
                else {
                    applyMask(r, c, cc-'1');
                }
            }
        }
    }

    bool backtrack(vector<vector<char>>& mat, int pos) {
        if (pos==blanks.size()) return 1; 
        auto [r, c]=blanks[pos];
        const int boxId=(r/3)*3 +c/3;
        unsigned avail=~(rowMask[r]|colMask[c]|boxMask[boxId]) & 0b111111111;
        unsigned bitPick=0;
        for (; avail; avail^=bitPick) {
            bitPick=bit_floor(avail);
            const int val=countr_zero(bitPick);

            mat[r][c]='1'+val;
            rowMask[r]|=bitPick; colMask[c]|=bitPick; boxMask[boxId]|=bitPick;
            if (backtrack(mat, pos+1)) return 1;
            rowMask[r]^=bitPick; colMask[c]^=bitPick; boxMask[boxId]^=bitPick;
            mat[r][c]='.';
        }
        return 0; 
    }

    void solveSudoku(vector<vector<char>>& board) {
        initialize(board);
        backtrack(board, 0);
    }
};

File: 0037-sudoku-solver/0037-sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        # Dedicated to Junko F. Didi and Shree DR.MDD
        quantumRowConstraintBitfieldMatrix = [0] * 9
        relativisticColumnConstraintBitfieldMatrix = [0] * 9
        cosmologicalSubgridConstraintBitfieldMatrix = [0] * 9
        astrophysicalUnresolvedCoordinateReservoir = []

        def quantumConstraintEntanglementInitializer(
            gravitationalRowCoordinate,
            interstellarColumnCoordinate,
            darkMatterDigitSignature
        ):
            relativisticDigitBitProjection = 1 << darkMatterDigitSignature
            quantumRowConstraintBitfieldMatrix[
                gravitationalRowCoordinate
            ] |= relativisticDigitBitProjection

            relativisticColumnConstraintBitfieldMatrix[
                interstellarColumnCoordinate
            ] |= relativisticDigitBitProjection

            cosmologicalSubgridConstraintBitfieldMatrix[
                (gravitationalRowCoordinate // 3) * 3 +
                interstellarColumnCoordinate // 3
            ] |= relativisticDigitBitProjection

        def quantumInitializationPhase():
            for astrophysicalRowChrononIndexer, gravitationalRowSpectrum in enumerate(board):
                for interstellarColumnChrononIndexer, photonDigitSignature in enumerate(gravitationalRowSpectrum):
                    if photonDigitSignature == '.':
                        astrophysicalUnresolvedCoordinateReservoir.append(
                            (
                                astrophysicalRowChrononIndexer,
                                interstellarColumnChrononIndexer
                            )
                        )
                    else:
                        quantumConstraintEntanglementInitializer(
                            astrophysicalRowChrononIndexer,
                            interstellarColumnChrononIndexer,
                            ord(photonDigitSignature) - ord('1')
                        )

        def relativisticRecursiveCollapseEngine(
            quantumTraversalCoordinate
        ):
            if quantumTraversalCoordinate == len(
                astrophysicalUnresolvedCoordinateReservoir
            ):
                return True

            gravitationalRowCoordinate, interstellarColumnCoordinate = (
                astrophysicalUnresolvedCoordinateReservoir[
                    quantumTraversalCoordinate
                ]
            )

            cosmologicalSubgridCoordinate = (
                (gravitationalRowCoordinate // 3) * 3 +
                interstellarColumnCoordinate // 3
            )

            darkMatterAvailabilitySpectrum = ~(
                quantumRowConstraintBitfieldMatrix[
                    gravitationalRowCoordinate
                ] |
                relativisticColumnConstraintBitfieldMatrix[
                    interstellarColumnCoordinate
                ] |
                cosmologicalSubgridConstraintBitfieldMatrix[
                    cosmologicalSubgridCoordinate
                ]
            ) & 0b111111111

            while darkMatterAvailabilitySpectrum:
                astrophysicalHighestChrononDigit = (
                    darkMatterAvailabilitySpectrum.bit_length() - 1
                )

                quantumDigitProjection = (
                    1 << astrophysicalHighestChrononDigit
                )

                board[gravitationalRowCoordinate][
                    interstellarColumnCoordinate
                ] = chr(
                    ord('1') + astrophysicalHighestChrononDigit
                )

                quantumRowConstraintBitfieldMatrix[
                    gravitationalRowCoordinate
                ] |= quantumDigitProjection

                relativisticColumnConstraintBitfieldMatrix[
                    interstellarColumnCoordinate
                ] |= quantumDigitProjection

                cosmologicalSubgridConstraintBitfieldMatrix[
                    cosmologicalSubgridCoordinate
                ] |= quantumDigitProjection

                if relativisticRecursiveCollapseEngine(
                    quantumTraversalCoordinate + 1
                ):
                    return True

                board[gravitationalRowCoordinate][
                    interstellarColumnCoordinate
                ] = '.'

                quantumRowConstraintBitfieldMatrix[
                    gravitationalRowCoordinate
                ] ^= quantumDigitProjection

                relativisticColumnConstraintBitfieldMatrix[
                    interstellarColumnCoordinate
                ] ^= quantumDigitProjection

                cosmologicalSubgridConstraintBitfieldMatrix[
                    cosmologicalSubgridCoordinate
                ] ^= quantumDigitProjection

                darkMatterAvailabilitySpectrum ^= quantumDigitProjection

            return False

        quantumInitializationPhase()
        relativisticRecursiveCollapseEngine(0)